File: 310971/Stankevich/redis350/publisher/src/main/java/com/task/rest/service/MarkService.java
package com.task.rest.service;

import com.task.rest.dto.MarkRequestTo;
import com.task.rest.dto.MarkResponseTo;
import com.task.rest.exception.ResourceNotFoundException;
import com.task.rest.exception.ValidationException;
import com.task.rest.mapper.MarkMapper;
import com.task.rest.model.Mark;
import com.task.rest.repository.MarkRepository;
import com.task.rest.service.cache.CacheService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional
public class MarkService {

    private final MarkRepository markRepository;
    private final MarkMapper markMapper;
    private final CacheService cacheService;

    public MarkResponseTo getById(Long id) {
        log.info("Getting mark by id: {}", id);

        // Try to get from cache first
        Optional<MarkResponseTo> cached = cacheService.getMarkFromCache(id, MarkResponseTo.class);
        if (cached.isPresent()) {
            log.debug("Mark found in cache: {}", id);
            return cached.get();
        }

        // Get from database
        Mark mark = markRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Mark not found with id: {}", id);
                    return new ResourceNotFoundException("Mark not found with id: " + id);
                });

        MarkResponseTo response = markMapper.toResponseTo(mark);

        // Cache the result
        cacheService.cacheMark(id, response);

        return response;
    }

    public List<MarkResponseTo> getAllList() {
        log.info("Getting all marks as list");
        return markRepository.findAll().stream()
                .map(markMapper::toResponseTo)
                .collect(Collectors.toList());
    }

    public MarkResponseTo create(MarkRequestTo requestTo) {
        log.info("Creating mark with name: {}", requestTo.getName());

        if (markRepository.existsByName(requestTo.getName())) {
            log.error("Mark with name already exists: {}", requestTo.getName());
            throw new ValidationException("Mark with name '" + requestTo.getName() + "' already exists");
        }

        Mark mark = markMapper.toEntity(requestTo);
        Mark saved = markRepository.save(mark);
        log.info("Mark created successfully with id: {}", saved.getId());

        MarkResponseTo response = markMapper.toResponseTo(saved);

        // Cache the new mark
        cacheService.cacheMark(saved.getId(), response);

        return response;
    }

    public MarkResponseTo update(Long id, MarkRequestTo requestTo) {
        log.info("Updating mark with id: {}", id);

        Mark mark = markRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Mark not found with id: {}", id);
                    return new ResourceNotFoundException("Mark not found with id: " + id);
                });

        if (!mark.getName().equals(requestTo.getName()) &&
                markRepository.existsByName(requestTo.getName())) {
            log.error("Mark with name already exists: {}", requestTo.getName());
            throw new ValidationException("Mark with name '" + requestTo.getName() + "' already exists");
        }

        mark.setName(requestTo.getName());

        Mark updated = markRepository.save(mark);
        log.info("Mark updated successfully with id: {}", id);

        MarkResponseTo response = markMapper.toResponseTo(updated);

        // Update cache
        cacheService.cacheMark(id, response);

        return response;
    }

    public void delete(Long id) {
        log.info("Deleting mark with id: {}", id);

        if (!markRepository.existsById(id)) {
            log.error("Mark not found with id: {}", id);
            throw new ResourceNotFoundException("Mark not found with id: " + id);
        }

        markRepository.deleteById(id);

        // Evict from cache
        cacheService.evictMark(id);

        log.info("Mark deleted successfully with id: {}", id);
    }
}

File: 351003/Pulish/publisher/app/services/mark_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from app.dto.mark import MarkRequestTo, MarkResponseTo
from app.models.mark import Mark
from app.core.exceptions import NotFoundException, AppException


class MarkService:
    def __init__(self, db: Session):
        self.db = db

    def create(self, dto: MarkRequestTo) -> MarkResponseTo:
        mark = Mark(name=dto.name)
        try:
            self.db.add(mark)
            self.db.commit()
            self.db.refresh(mark)
        except IntegrityError:
            self.db.rollback()
            raise AppException(
                "Mark with this name already exists", 40303, 403)
        return self._to_response(mark)

    def find_all(self):
        marks = self.db.query(Mark).all()
        return [self._to_response(m) for m in marks]

    def find_by_id(self, id: int):
        mark = self.db.query(Mark).filter(Mark.id == id).first()
        if not mark:
            raise NotFoundException("Mark not found", 40403)
        return self._to_response(mark)

    def update(self, dto: MarkRequestTo):
        mark = self.db.query(Mark).filter(Mark.id == dto.id).first()
        if not mark:
            raise NotFoundException("Mark not found", 40403)

        mark.name = dto.name
        try:
            self.db.commit()
            self.db.refresh(mark)
        except IntegrityError:
            self.db.rollback()
            raise AppException(
                "Mark with this name already exists", 40303, 403)

        return self._to_response(mark)

    def delete(self, id: int):
        mark = self.db.query(Mark).filter(Mark.id == id).first()
        if not mark:
            raise NotFoundException("Mark not found", 40403)
        self.db.delete(mark)
        self.db.commit()

    def _to_response(self, mark: Mark) -> MarkResponseTo:
        return MarkResponseTo(id=mark.id, name=mark.name)
